### DISCUSSION TITLE
What is the difference between  *>*, *+* , * * * and body

### DISCUSSION THREAD

**Discussion by jacobovillegas** [cssbattle/forum#190] [Q&A] [answered]
If the title is a little confusing here is with better format.

What's the difference between:

`*>*`
`*>*`
`* * *`
`body`

I have a vague idea of the first two ones and the obvious `*` and i am guilty of interchanging between `*>*` and `*+*` but i do that based on trial and error. I would like to know why some of my solutions work with `body` and not with `*>*`

ps: oh also i have a question that's been in my head for months, one of my solutions (don't remember which) works with `background:#` just like that, without Hex code number o_o

**Comment by Beowolve** [comment]
@SelenIT explained it here already.
https://github.com/cssbattle/forum/discussions/154#discussioncomment-1583056

**Comment by rafehimself** [comment]
there is also `* *` and `*~*`

**Reply by jacobovillegas** [reply]
`* *` is the same as `*+*` if i'm not wrong. You can replace almost any space between characters with a `+`. And `*~*`, never heard of that, do you know what represents?

**Comment by SelenIT** [comment, marked as answer]
It's basically [CSS combinators](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Selectors/Combinators):

`* *` is anything that is anything's _descendant_
`*>*` is anything that is anything's _child_ (immediate descendant)
`* * *` is anything that has two or more ancestors, or is anything's "grandchild"
`*+*` is anything that is anything's next sibling, or has a previous sibling
`*~*` is anything that is anything's following sibling, or has some preceding siblings (at least one)

With universal selectors, there is no practical difference between `* *` and `*>*`, or between `*+*` and `*~*`. However, they can be different with other selectors (`div div` will match the inner div in  `<div><section><div></div></section></div>` while `div > div` won't. `dd ~ dd` will match the second `<dd>` in `<dd></dd><dt></dt><dd></dd>` while `dd + dd` won't).

In the empty HTML page. `* *` and `*>*` will match everything except HTML, i.e. `<head>` (invisible by default and thus irrelevant) and `<body>`. while `*+*` will select only `<body>` (as the `<head>`'s next sibling). In the nonempty HTML page, `* * *` will select all descendants of `<body>`, but not `<html>` or `<body>` itself.

In the page starting with  `<p><p><style>...`, resulting in the following DOM structure:
```HTML
<html>
  <head></head>
  <body>
    <p></p>
    <p><style>...</style></p>
  </body>
</html>
```
`* *` and `*>*` will match `<head>`, `<body>`, both `<p>`s, and `<style>` (with `<head>` and `<style>` invisible by default. but it can change with `display` setting);
`*+*` and `*~*` will match `<body>` and the second `<p>` (but not the 1st `<p>`!);
`* * *`. `*>*>*`, `* *>*`, and `*>* *` will match both `<p>`s (and invisible `<style>`);
`*+* *` also will match both `<p>`s and invisible `<style>` (as descendants of `body`);
`*+*>*` also will match both `<p>`s (as direct children of `<body>`) and `<style>` (as direct child of the second `p`);
`* *+*` and `*>*+*` also will match `<body>` and the second `<p>` (as the following siblings of an element that has an ancestor/parent);
`* * *+*`, `*>*>*+*`, `* *>*+*`, and `*>* *+*` will match only the second `<p>` (so better use `p+p`, `*+p`, or `p+*` instead);
`*+*+*` will match nothing (there is no element with 2 preceding siblings in this structure).

Also, any combination of `*`s has the same zero specificity (lower than tag selector specificity), so the following:
```CSS
*+*{
  /* intended for BODY */
  background: red;
}
*{
  /* intended for HTML only */
  background: radial-gradient(...
```
will not work — the declaration for `*` will override the declaration for `*+*`, so you will need either to swap the rules making the `*+*` selector come later (and losing the abilitly to drop the closing `)` of gradient) or to replace `*+*` with `body`. Both ways result in the same code length. However, the following:
```CSS
*+*{
  --b: red;
}
*{
  background: var(--b,radial-gradient(...
```
will work. since the custom property will be defined for `body` only, so only `html` will get the fallback value.

**Reply by jacobovillegas** [reply]
As always, great answer. Thank you.

**Reply by SelenIT** [reply]
Sorry, there was a mistake in previous version of my reply regarding the `* *+*` selector and the like 😳. Now it's corrected.
